src/layout/umap.c

```diff
@@ -0,0 +1,1038 @@
+/* -*- mode: C -*-  */
+/*
+   IGraph library.
+   Copyright (C) 2008-2021  The igraph development team <igraph@igraph.org>
+
+   This program is free software; you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation; either version 2 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+   */
+
+#include "igraph_constructors.h"
+#include "igraph_interface.h"
+#include "igraph_lapack.h"
+#include "igraph_layout.h"
+#include "igraph_matrix.h"
+#include "igraph_random.h"
+#include "igraph_nongraph.h"
+
+
+/* rho is just the size of the distance from each vertex and its closest neighbor */
+/* sigma is the the decay from each vertex, depends on its rho and the rest of its neighbor
+ * distances */
+
+/* Find sigma for this vertex by binary search */
+static igraph_error_t igraph_i_umap_find_sigma(const igraph_t *graph,
+        const igraph_vector_t *distances, igraph_integer_t i, const igraph_vector_int_t *eids,
+        igraph_real_t rho, igraph_real_t *sigma_p, igraph_real_t target) {
+
+    igraph_real_t sigma = 1;
+    igraph_real_t sum;
+    igraph_real_t tol = 0.01;
+    igraph_integer_t maxiter = 100;
+    igraph_integer_t no_of_neis = igraph_vector_int_size(eids);
+    igraph_integer_t eid;
+    igraph_real_t step = sigma;
+    igraph_integer_t seen_max = 0;
+
+    /* Binary search */
+    for (igraph_integer_t iter = 0; iter < maxiter; iter++) {
+        sum = 0;
+        for (igraph_integer_t j = 0; j < no_of_neis; j++) {
+            eid = VECTOR(*eids)[j];
+            sum += exp(-(VECTOR(*distances)[eid] - rho) / sigma);
+        }
+
+#ifdef UMAP_DEBUG
+        printf("SIGMA function (i = %" IGRAPH_PRId ", no_of_neis = %" IGRAPH_PRId ")- sum: %g, "
+               "target: %g, rho: %g, sigma: %g\n", i, no_of_neis, sum, target, rho, sigma);
+#endif
+
+        /* TODO: this seems fine, but is probably a little off for some corner cases */
```

Comment:
If someone other than you comes back to this in the future, they will not be able to tell what this comment is about.

Reply:
yep, this is done